### ISSUE TITLE
"wall_clock_breakdown": true and overlap_comm: true

### ISSUE THREAD

**Issue by szhengac** [microsoft/DeepSpeed#617]
Hi, I have a question regarding how deepspeed measures the communication time. I see that there is a timer that counts the time for allreduce as in https://github.com/microsoft/DeepSpeed/blob/81aeea361da3936b875a678b9cb44596800510b5/deepspeed/runtime/engine.py#L865-L874. But when I go into this function, I found it eventually goes into https://github.com/microsoft/DeepSpeed/blob/81aeea361da3936b875a678b9cb44596800510b5/deepspeed/runtime/zero/stage2.py#L468-L509, which only performs cuda synchronization when `overlap_comm=True`. If I remember correctly, pytorch backward is a blocking operation, and the communication finishes before we enter self.timers('backward_allreduce_microstep').start() .

**Comment by tjruwase** [comment]
@szhengac Your analysis is correct. Communication overlapping breaks correctness of the allreduce timer.

**Comment by szhengac** [comment]
@tjruwase Thanks for confirming this. And, I have one more question, I see that `reduce_ipg_grads` in https://github.com/microsoft/DeepSpeed/blob/81aeea361da3936b875a678b9cb44596800510b5/deepspeed/runtime/zero/stage2.py#L942 is called both inside the gradient hook (https://github.com/microsoft/DeepSpeed/blob/81aeea361da3936b875a678b9cb44596800510b5/deepspeed/runtime/zero/stage2.py#L593-L596) and after optimizer.backward() https://github.com/microsoft/DeepSpeed/blob/81aeea361da3936b875a678b9cb44596800510b5/deepspeed/runtime/zero/stage2.py#L470 when "overlap_comm=True". It seems to me we are performing allreduce twice?

**Comment by szhengac** [comment]
I think I find the idea here. We add hook to the gradient no matter if `overlap_comm=True` or `False`. The overlap is enabled by using two cuda streams?

**Comment by tjruwase** [comment]
Yes, your idea is on the right path. In ZeRO stage 2, we enable reduction doing backward after a predefined [number](https://github.com/microsoft/DeepSpeed/blob/81aeea361da3936b875a678b9cb44596800510b5/deepspeed/runtime/zero/stage2.py#L614) of gradients are available.  The number of gradients to trigger reduce is controlled by [reduce_bucket_size](https://www.deepspeed.ai/docs/config-json/#zero-optimizations-for-fp16-training) in the json config. As you have observed `overlap_comm=True` allows this reduction to overlap the remaining backward operations. The extra reduce called in engine.py at the end of backward will reduce any remaining gradients (e.g., if gradient count is not multiple of bucket size), in addition to doing postprocessing. Hope that helps.

**Comment by szhengac** [comment]
Thanks. This explains. I think it may be helpful to add some comments around that allreduce call, otherwise it is misleading. And, since deepspeed uses its own launcher, I cannot directly just use `nsys --profile deepspeed <args>` for profiling.  mpirun https://github.com/microsoft/DeepSpeed/issues/461 also has issue. Is there a way that I can use Nsight to profile the job?

**Comment by szhengac** [comment]
Nevermind, I have managed to fix the mpirun and used nsight to profile the training. I put the profiling result in a new issue #620.

**Comment by kehuanfeng** [comment]
> Nevermind, I have managed to fix the mpirun and used nsight to profile the training. I put the profiling result in a new issue #620.

@szhengac could you please share how you get nsight working with deepspeed launcher?
The command I am using is as below, and the report doesn't contain any cuda trace information.
`nsys profile --trace=cuda deepspeed`
